{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Mummer pipeline \n",
    "Commands for running promer/nucmer for each _Drosophila_ species\n",
    "\n",
    "**Versions**  \n",
    "Mummer: 4.0.0beta2  \n",
    "Python: 3.7.3  \n",
    "GNUPLOT: 4.6.0  \n",
    "ghostscript: 9.22\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a muller element reference for _D. melanogaster_     \n",
    "Reference: /projects/genetics/ellison_lab/genomes/dna/dmel-all-chromosome-r6.22.fasta\n",
    "\n",
    "After running extract_scaff_rename_muller --  \n",
    "dmel-all-chromosome-r6.22_extract_ABCDEF_renamed.fasta **(This file used for downstream analysis)**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**_D. ananassae_**  \n",
    "Reference: dana_pacbio_GCA_003285975.fa  \n",
    "Final files: <tt> /projects/genetics/ellison_lab/nicole/mummer_projects/species/dana_new/ <tt>\n",
    "\n",
    "|Muller element | Scaffold |\n",
    "|--- | ---\n",
    "Muller_ A|HiC_Scaffold_3|\n",
    "Muller_B |HiC_scaffold_5|\n",
    "Muller_C |HiC_scaffold_4|\n",
    "Muller_D |HiC_scaffold_7|\n",
    "Muller_E |HiC_scaffold_6|\n",
    "Muller_F |HiC_scaffold_2|"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "python extract_scaff_rename_muller_plus_other_scaffs.py -i dana_pacbio_GCA_003285975.FINAL.fasta -a HiC_scaffold_3 --rcMB HiC_scaffold_5 --rcMC HiC_scaffold_4 --rcMD HiC_scaffold_7 -e HiC_scaffold_6 -f HiC_scaffold_2 --o2 dana_muller_o.fasta -s dana_other_scaffolds.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "#SBATCH --partition=main    # Partition (job queue)\n",
    "#SBATCH --job-name=mummer         # Assign an 8-character name to your job, no spaces\n",
    "#SBATCH --nodes=1                # Number of compute nodes\n",
    "#SBATCH --ntasks=1               # Processes (usually = cores) on each node\n",
    "#SBATCH --cpus-per-task=28       # Threads per process (or per core)\n",
    "#SBATCH --export=ALL             # Export you current environment settings to the job environment\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH --mem=20G\n",
    "#SBATCH --output=/scratch/nt365/mummer/mummer_projects/dana_new/slurm-%A_%a.out\n",
    "\n",
    "source activate MUMMER\n",
    "promer --mum -b 10 -l 50 --coords -p promer_mum_dana_rcMCMD /home/nt365/projects/mummer/script/dmel-all-chromosome-r6.22_extract_ABCDEF.fasta dana_muller_o.fasta\n",
    "\n",
    "mummerplot -p promer_mum_dana_rcMCMD -t postscript promer_mum_dana_rcMCMD.delta\n",
    "\n",
    "export GNUPLOT_PS_DIR=/home/nt365/pkg/gnuplot-4.6.0/term/PostScript \n",
    "gnuplot promer_mum_dana_rcMCMD.gp\n",
    "ps2pdf promer_mum_dana_rcMCMD.ps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**_D.biarmipes_**  \n",
    "Reference: Dbia_r2_illumina_patched.fa  \n",
    "Final files: <tt>/projects/genetics/ellison_lab/nicole/mummer_projects/species/dmel_dbia/<tt>\n",
    "\n",
    "|Muller element | Scaffold |\n",
    "|--- | ---\n",
    "Muller_ A|HiC_Scaffold_10|\n",
    "Muller_B |HiC_scaffold_7|\n",
    "Muller_C |HiC_scaffold_6|\n",
    "Muller_D |HiC_scaffold_3|\n",
    "Muller_E |HiC_scaffold_2|\n",
    "Muller_F |HiC_scaffold_9|"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "python extract_scaff_rename_muller.py -i Dbia_r2_illumina_patched.scaffolds.fasta -o dbia_muller_o.fasta -t dbia_muller_o.fasta -a HiC_scaffold_10 -b HiC_scaffold_7 -c HiC_scaffold_6 -d HiC_scaffold_3 -e HiC_scaffold_2 -f HiC_scaffold_9 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "#SBATCH --partition=main    # Partition (job queue)\n",
    "#SBATCH --job-name=mummer         # Assign an 8-character name to your job, no spaces\n",
    "#SBATCH --nodes=1                # Number of compute nodes\n",
    "#SBATCH --ntasks=1               # Processes (usually = cores) on each node\n",
    "#SBATCH --cpus-per-task=28       # Threads per process (or per core)\n",
    "#SBATCH --export=ALL             # Export you current environment settings to the job environment\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH --mem=20G\n",
    "#SBATCH --output=/scratch/nt365/mummer/mummer_projects/dana_new2/slurm-%A_%a.out\n",
    "\n",
    "source activate MUMMER\n",
    "promer --mum -b 10 -l 50 --coords -p promer_mum_dana_rcMBMCMD /home/nt365/projects/mummer/script/dmel-all-chromosome-r6.22_extract_ABCDEF.fasta dana_muller_o.fasta\n",
    "\n",
    "mummerplot -p promer_mum_dana_rcMBMCMD -t postscript promer_mum_dana_rcMBMCMD.delta\n",
    "\n",
    "export GNUPLOT_PS_DIR=/home/nt365/pkg/gnuplot-4.6.0/term/PostScript \n",
    "gnuplot promer_mum_dana_rcMBMCMD.gp\n",
    "ps2pdf promer_mum_dana_rcMBMCMD.ps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**_D. eugracilis_**   \n",
    "Reference: Deug_r2_illumina_patched.fa  \n",
    "Final files: /projects/genetics/ellison_lab/nicole/mummer_projects/species/dmel_deug\n",
    "\n",
    "|Muller element | Scaffold |\n",
    "|--- | ---\n",
    "Muller_ A|HiC_Scaffold_4|\n",
    "Muller_B |HiC_scaffold_6|\n",
    "Muller_C |HiC_scaffold_7|\n",
    "Muller_D |HiC_scaffold_13|\n",
    "Muller_E |HiC_scaffold_12|\n",
    "Muller_F |HiC_scaffold_10|"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "python extract_scaff_rename_muller.py -i Deug_r2_illumina_patched.scaffolds.fasta -o deug_muller_o.fasta -t deug_muller_o.fasta -a HiC_scaffold_4 -b HiC_scaffold_6 -c HiC_scaffold_7 -d HiC_scaffold_13 -e HiC_scaffold_12 -f HiC_scaffold_10 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "#SBATCH --partition=main    # Partition (job queue)\n",
    "#SBATCH --job-name=mummer         # Assign an 8-character name to your job, no spaces\n",
    "#SBATCH --nodes=1                # Number of compute nodes\n",
    "#SBATCH --ntasks=1               # Processes (usually = cores) on each node\n",
    "#SBATCH --cpus-per-task=28       # Threads per process (or per core)\n",
    "#SBATCH --export=ALL             # Export you current environment settings to the job environment\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH --mem=20G\n",
    "#SBATCH --output=/scratch/nt365/mummer/mummer_projects/dmel_deug/slurm-%A_%a.out\n",
    "\n",
    "source activate MUMMER\n",
    "nucmer --mum -t 3 -c 400 -L 400 -b 100 -p nucmer_mum_deug /home/nt365/projects/mummer/script/dmel-all-chromosome-r6.22_extract_ABCDEF.fasta deug_muller_ordered.fasta\n",
    "\n",
    "mummerplot -p nucmer_mum_deug -t postscript nucmer_mum_deug.delta\n",
    "\n",
    "export GNUPLOT_PS_DIR=/home/nt365/pkg/gnuplot-4.6.0/term/PostScript \n",
    "gnuplot nucmer_mum_deug.gp\n",
    "ps2pdf nucmer_mum_deug.ps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**_D. fic_**   \n",
    "\n",
    "Final files: <tt>/projects/genetics/ellison_lab/nicole/mummer_projects/species/dmel_dfic<tt>\n",
    "\n",
    "|Muller element | Scaffold |\n",
    "|--- | ---\n",
    "Muller_ A|HiC_Scaffold_6|\n",
    "Muller_B |HiC_scaffold_1|\n",
    "Muller_C |HiC_scaffold_13|\n",
    "Muller_D |HiC_scaffold_5|\n",
    "Muller_E |HiC_scaffold_11|\n",
    "Muller_F |HiC_scaffold_7|"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "python extract_scaff_rename_muller.py -i Dfic_r2_illumina.reformat.scaffolds.fasta -o dfic_muller_o.fasta -t dfic_muller_o.fasta -a HiC_scaffold_6 -b HiC_scaffold_1 --rcMC HiC_scaffold_13 --rcMD HiC_scaffold_5 -e HiC_scaffold_11 -f HiC_scaffold_7 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "#SBATCH --partition=main    # Partition (job queue)\n",
    "#SBATCH --job-name=mummer         # Assign an 8-character name to your job, no spaces\n",
    "#SBATCH --nodes=1                # Number of compute nodes\n",
    "#SBATCH --ntasks=1               # Processes (usually = cores) on each node\n",
    "#SBATCH --cpus-per-task=28       # Threads per process (or per core)\n",
    "#SBATCH --export=ALL             # Export you current environment settings to the job environment\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH --mem=20G\n",
    "#SBATCH --output=/scratch/nt365/mummer/mummer_projects/dmel_dfic/slurm-%A_%a.out\n",
    "\n",
    "source activate MUMMER\n",
    "promer --mum -b 10 -l 50 --coords -p promer_mum_dfic_rcMCMD /home/nt365/projects/mummer/script/dmel-all-chromosome-r6.22_extract_ABCDEF.fasta dfic_muller_o.fasta\n",
    "\n",
    "mummerplot -p promer_mum_dfic_rcMCMD -t postscript promer_mum_dfic_rcMCMD.delta\n",
    "\n",
    "export GNUPLOT_PS_DIR=/home/nt365/pkg/gnuplot-4.6.0/term/PostScript \n",
    "gnuplot promer_mum_dfic_rcMCMD.gp\n",
    "ps2pdf promer_mum_dfic_rcMCMD.ps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**_D.takahashii_**   \n",
    "Reference: Dtak_r2_illumina_patched.fa   \n",
    "Final files: <tt>/projects/genetics/ellison_lab/nicole/mummer_projects/species/dmel_dtak/<tt>\n",
    "\n",
    "|Muller element | Scaffold |\n",
    "|--- | ---\n",
    "Muller_ A|HiC_Scaffold_2|\n",
    "Muller_B |HiC_scaffold_6|\n",
    "Muller_C |HiC_scaffold_7|\n",
    "Muller_D |HiC_scaffold_10|\n",
    "Muller_E |HiC_scaffold_4|\n",
    "Muller_F |HiC_scaffold_8|"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "python extract_scaff_rename_muller.py -i Dtak_r2_illumina.reformat.scaffolds.fasta -o dtak_muller_o.fasta -t dtak_muller_o.fasta --rcMA HiC_scaffold_2 --rcMB HiC_scaffold_6 --rcMC HiC_scaffold_7 --rcMD HiC_scaffold_10 --rcME HiC_scaffold_4 --rcMF HiC_scaffold_8 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "#SBATCH --partition=main    # Partition (job queue)\n",
    "#SBATCH --job-name=mummer         # Assign an 8-character name to your job, no spaces\n",
    "#SBATCH --nodes=1                # Number of compute nodes\n",
    "#SBATCH --ntasks=1               # Processes (usually = cores) on each node\n",
    "#SBATCH --cpus-per-task=28       # Threads per process (or per core)\n",
    "#SBATCH --export=ALL             # Export you current environment settings to the job environment\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH --mem=20G\n",
    "#SBATCH --output=/scratch/nt365/mummer/mummer_projects/dmel_dtak/slurm-%A_%a.out\n",
    "\n",
    "source activate MUMMER\n",
    "promer --mum -b 10 -l 50 --coords -p promer_mum_dtak /home/nt365/projects/mummer/script/dmel-all-chromosome-r6.22_extract_ABCDEF.fasta dtak_muller_o.fasta\n",
    "\n",
    "mummerplot -p promer_mum_dtak -t postscript promer_mum_dtak.delta\n",
    "\n",
    "export GNUPLOT_PS_DIR=/home/nt365/pkg/gnuplot-4.6.0/term/PostScript \n",
    "gnuplot promer_mum_dtak.gp\n",
    "ps2pdf promer_mum_dtak.ps\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**_D.yakuba_**  \n",
    "Reference: Dyak.pass.minimap2.racon.x3.pilon.x3.fasta  \n",
    "Final files: <tt>/projects/genetics/ellison_lab/nicole/mummer_projects/species/dmel_dyak/<tt>\n",
    "\n",
    "|Muller element | Scaffold |\n",
    "|--- | ---\n",
    "Muller_A |HiC_Scaffold_1|\n",
    "Muller_B |HiC_scaffold_5|\n",
    "Muller_C |HiC_scaffold_6|\n",
    "Muller_D |HiC_scaffold_4|\n",
    "Muller_E |HiC_scaffold_3|\n",
    "Muller_F |HiC_scaffold_2|"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "python extract_scaff_rename_muller.py -i Dyak.pass.minimap2.racon.x3.pilon.x3.FINAL.fasta -o dyak_muller_o.fasta --o2 dyak_muller_o.fasta -a HiC_scaffold_1 -b HiC_scaffold_5 -c HiC_scaffold_6 --rcMD HiC_scaffold_4 --rcME HiC_scaffold_3 -f HiC_scaffold_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "#SBATCH --partition=main    # Partition (job queue)\n",
    "#SBATCH --job-name=mummer         # Assign an 8-character name to your job, no spaces\n",
    "#SBATCH --nodes=1                # Number of compute nodes\n",
    "#SBATCH --ntasks=1               # Processes (usually = cores) on each node\n",
    "#SBATCH --cpus-per-task=28       # Threads per process (or per core)\n",
    "#SBATCH --export=ALL             # Export you current environment settings to the job environment\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH --mem=20G\n",
    "#SBATCH --output=/scratch/nt365/mummer/mummer_projects/dmel_dyak/slurm-%A_%a.out\n",
    "\n",
    "nucmer --mum -t 3 -c 400 -L 400 -b 100 -p nucmer_mum_dyak_rcMCMDME /home/nt365/projects/mummer/script/dmel-all-chromosome-r6.22_extract_ABCDEF.fasta dyak_muller_o.fasta\n",
    "\n",
    "mummerplot -p nucmer_mum_dyak_rcMCMDME -t postscript nucmer_mum_dyak_rcMCMDME.delta\n",
    "\n",
    "export GNUPLOT_PS_DIR=/home/nt365/pkg/gnuplot-4.6.0/term/PostScript \n",
    "gnuplot nucmer_mum_dyak_rcMCMDME.gp\n",
    "ps2pdf nucmer_mum_dyak_rcMCMDME.ps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**_D. elegans_**  \n",
    "Reference: Dele_r2_illumina_patched.fa   \n",
    "Final files: <tt>/projects/genetics/ellison_lab/nicole/mummer_projects/species/dmel_dele/<tt>\n",
    "\n",
    "|Muller element | Scaffold |\n",
    "|--- | ---\n",
    "Muller_ A|HiC_Scaffold_4|\n",
    "Muller_B |HiC_scaffold_11|\n",
    "Muller_C |HiC_scaffold_3|\n",
    "Muller_D |HiC_scaffold_1|\n",
    "Muller_E |HiC_scaffold_10|\n",
    "Muller_F |HiC_scaffold_6|"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "python extract_scaff_rename_muller.py -i Dele_r2_illumina_patched.scaffolds.fasta -o dele_renamed.fasta -t dele_muller_o.fasta -a HiC_scaffold_4 --rcMB HiC_scaffold_11 -c HiC_scaffold_3 -d HiC_scaffold_1 -e HiC_scaffold_10 --rcMF HiC_scaffold_6 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "#SBATCH --partition=main    # Partition (job queue)\n",
    "#SBATCH --job-name=mummer         # Assign an 8-character name to your job, no spaces\n",
    "#SBATCH --nodes=1                # Number of compute nodes\n",
    "#SBATCH --ntasks=1               # Processes (usually = cores) on each node\n",
    "#SBATCH --cpus-per-task=28       # Threads per process (or per core)\n",
    "#SBATCH --export=ALL             # Export you current environment settings to the job environment\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH --mem=20G\n",
    "#SBATCH --output=/scratch/nt365/mummer/mummer_projects/dmel_dele/slurm-%A_%a.out\n",
    "\n",
    "source activate MUMMER\n",
    "promer --mum -b 10 -l 50 --coords -p promer_mum_dele_rcMBMF /home/nt365/projects/mummer/script/dmel-all-chromosome-r6.22_extract_ABCDEF.fasta dele_muller_o.fasta\n",
    "\n",
    "mummerplot -p promer_mum_dele_rcMBMF -t postscript promer_mum_dele_rcMBMF.delta\n",
    "\n",
    "export GNUPLOT_PS_DIR=/home/nt365/pkg/gnuplot-4.6.0/term/PostScript \n",
    "gnuplot promer_mum_dele_rcMBMF.gp\n",
    "ps2pdf promer_mum_dele_rcMBMF.ps\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**_D. triauraria_**  \n",
    "Reference: tria_20181107_r2_illumina.reformat.scaffolds.fasta\n",
    "Final files: <tt>/projects/genetics/ellison_lab/nicole/mummer_projects/species/dmel_dtria/<tt>\n",
    "\n",
    "|Muller element | Scaffold |\n",
    "|--- | ---\n",
    "Muller_ A|HiC_Scaffold_5|\n",
    "Muller_B |HiC_scaffold_1|\n",
    "Muller_C |HiC_scaffold_2|\n",
    "Muller_D |HiC_scaffold_7|\n",
    "Muller_E |HiC_scaffold_8|\n",
    "Muller_F |HiC_scaffold_6|"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "python extract_scaff_rename_muller.py -i tria_20181107_r2_illumina.reformat.scaffolds.fasta -o dtria_muller_o.fasta -t dtria_muller_o.fasta --rcMA HiC_scaffold_5 --rcMB HiC_scaffold_1 -c HiC_scaffold_2 -d HiC_scaffold_7 -e HiC_scaffold_8 -f HiC_scaffold_6 \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "#SBATCH --partition=main    # Partition (job queue)\n",
    "#SBATCH --job-name=mummer         # Assign an 8-character name to your job, no spaces\n",
    "#SBATCH --nodes=1                # Number of compute nodes\n",
    "#SBATCH --ntasks=1               # Processes (usually = cores) on each node\n",
    "#SBATCH --cpus-per-task=28       # Threads per process (or per core)\n",
    "#SBATCH --export=ALL             # Export you current environment settings to the job environment\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH --mem=20G\n",
    "#SBATCH --output=/scratch/nt365/mummer/mummer_projects/dmel_dtria/slurm-%A_%a.out\n",
    "\n",
    "source activate MUMMER\n",
    "promer --mum -b 10 -l 50 --coords -p promer_mum_dtria_rcMAMB /home/nt365/projects/mummer/script/dmel-all-chromosome-r6.22_extract_ABCDEF.fasta dtria_muller_o.fasta\n",
    "\n",
    "mummerplot -p promer_mum_dtria_rcMAMB -t postscript promer_mum_dtria_rcMAMB.delta\n",
    "\n",
    "export GNUPLOT_PS_DIR=/home/nt365/pkg/gnuplot-4.6.0/term/PostScript \n",
    "gnuplot promer_mum_dtria_rcMAMB.gp\n",
    "ps2pdf promer_mum_dtria_rcMAMB.ps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**_D. erecta_**  \n",
    "Reference: dere_pacbio_GCA_003286155.scaffolds.fasta  \n",
    "Final files: <tt>/projects/genetics/ellison_lab/nicole/mummer_projects/species/dere_new/<tt>\n",
    "\n",
    "|Muller element | Scaffold |\n",
    "|--- | ---\n",
    "Muller_ A|HiC_Scaffold_2|\n",
    "Muller_B |HiC_scaffold_7|\n",
    "Muller_C |HiC_scaffold_6|\n",
    "Muller_D |HiC_scaffold_3|\n",
    "Muller_E |HiC_scaffold_4|\n",
    "Muller_F |HiC_scaffold_5|"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "python extract_scaff_rename_muller.py -i dere_pacbio_GCA_003286155.FINAL.fasta -o dere_muller_o.fasta --o2 dere_muller_o.fasta --rcMA HiC_scaffold_2 -b HiC_scaffold_7 -c HiC_scaffold_6 --rcMD HiC_scaffold_3 --rcME HiC_scaffold_4 --rcMF HiC_scaffold_5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "#SBATCH --partition=main    # Partition (job queue)\n",
    "#SBATCH --job-name=mummer         # Assign an 8-character name to your job, no spaces\n",
    "#SBATCH --nodes=1                # Number of compute nodes\n",
    "#SBATCH --ntasks=1               # Processes (usually = cores) on each node\n",
    "#SBATCH --cpus-per-task=28       # Threads per process (or per core)\n",
    "#SBATCH --export=ALL             # Export you current environment settings to the job environment\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH --mem=20G\n",
    "#SBATCH --output=/scratch/nt365/mummer/mummer_projects/dere_new/slurm-%A_%a.out\n",
    "\n",
    "source activate MUMMER\n",
    "nucmer --mum -t 3 -c 400 -L 400 -b 100 -p nucmer_mum_dere /home/nt365/projects/mummer/script/dmel-all-chromosome-r6.22_extract_ABCDEF.fasta dere_muller_o.fasta\n",
    "\n",
    "mummerplot -p nucmer_mum_dere -t postscript nucmer_mum_dere.delta\n",
    "\n",
    "export GNUPLOT_PS_DIR=/home/nt365/pkg/gnuplot-4.6.0/term/PostScript \n",
    "gnuplot nucmer_mum_dere.gp\n",
    "ps2pdf nucmer_mum_dere.ps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**_D. simulans_**  \n",
    "Reference: Dsim.pass.minimap2.racon.x3.pilon.x3.fasta\n",
    "Final files: <tt>/projects/genetics/ellison_lab/nicole/mummer_projects/species/dsim_new/<tt>\n",
    "\n",
    "|Muller element | Scaffold |\n",
    "|--- | ---\n",
    "Muller_A |HiC_Scaffold_1|\n",
    "Muller_B |HiC_scaffold_7|\n",
    "Muller_C |HiC_scaffold_6|\n",
    "Muller_D |HiC_scaffold_2|\n",
    "Muller_E |HiC_scaffold_4|\n",
    "Muller_F |HiC_scaffold_5|"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "python extract_scaff_rename_muller.py -i Dsim.pass.minimap2.racon.x3.pilon.x3.FINAL.fasta -o dsim_muller_o.fasta --o2 dsim_muller_o.fasta --rcMA HiC_scaffold_1 --rcMB HiC_scaffold_7 --rcMC HiC_scaffold_6 -d HiC_scaffold_2 --rcME HiC_scaffold_4 -f HiC_scaffold_5 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "#SBATCH --partition=main    # Partition (job queue)\n",
    "#SBATCH --job-name=mummer         # Assign an 8-character name to your job, no spaces\n",
    "#SBATCH --nodes=1                # Number of compute nodes\n",
    "#SBATCH --ntasks=1               # Processes (usually = cores) on each node\n",
    "#SBATCH --cpus-per-task=28       # Threads per process (or per core)\n",
    "#SBATCH --export=ALL             # Export you current environment settings to the job environment\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH --mem=20G\n",
    "#SBATCH --output=/scratch/nt365/mummer/mummer_projects/dmel_dsim_rcMAMBMCME/slurm-%A_%a.out\n",
    "\n",
    "source activate MUMMER\n",
    "nucmer --mum -t 3 -c 400 -L 400 -b 100 -p nucmer_mum_dsim_rcMAMBMCME /home/nt365/projects/mummer/script/dmel-all-chromosome-r6.22_extract_ABCDEF.fasta dsim_muller_o.fasta\n",
    "<tt>**/projects/genetics/ellison_lab/nicole/mummer_projects/script/**<tt>\n",
    "\n",
    "instructions for \"extract_scaff_rename_muller.py\"\n",
    "Script is to rename HiC scaffolds from Drosophila species to appropriate Muller elements and reverse complement the scaffold if necessary.\n",
    "\n",
    "Be sure to input two output files, the first will print the Muller elements in the order they are found and the second will be the reordered Muller elements (A,B,C,D,E,F).\n",
    "\n",
    "in bash shell type:\n",
    "python extract_scaff_rename_muller.py -h \n",
    "\n",
    "Output will be \n",
    "Options:\n",
    "  -h, --help            show this help message and exit\n",
    "  -i FASTA              scaffolds.fasta\n",
    "  -o OUTFILE            output1.fasta\n",
    "  -t OUTFILE2, --o2=OUTFILE2\n",
    "                        output2.fasta - this will be the correctly ordered\n",
    "                        file\n",
    "  -a MA                 scaffold corresp. to Muller A\n",
    "  -b MB                 scaffold corresp. to Muller B\n",
    "  -c MC                 scaffold corresp. to Muller C\n",
    "  -d MD                 scaffold corresp. to Muller D\n",
    "  -e ME                 scaffold corresp. to Muller E\n",
    "  -f MF                 scaffold corresp. to Muller F\n",
    "  -u MAR, --rcMA=MAR    scaffold corresp. to Muller A - use to reverse\n",
    "                        complement scaff\n",
    "  -v MBR, --rcMB=MBR    scaffold corresp. to Muller B - use to reverse\n",
    "                        complement scaff\n",
    "  -w MCR, --rcMC=MCR    scaffold corresp. to Muller C - use to reverse\n",
    "                        complement scaff\n",
    "  -x MDR, --rcMD=MDR    scaffold corresp. to Muller D - use to reverse\n",
    "                        complement scaff\n",
    "  -y MER, --rcME=MER    scaffold corresp. to Muller E - use to reverse\n",
    "                        complement scaff\n",
    "  -z MFR, --rcMF=MFR    scaffold corresp. to Muller F - use to reverse\n",
    "                        complement scaff\n",
    "\n",
    "\n",
    "After running the script, input OUTFILE2 name into the <OUTFILE2> slot in the nucmer_slurm.sh script in the folder. Update prefix for nucmer output files, the folder for the SLURM output, and the nucmer parameters as necessary.  \n",
    "\n",
    "mummerplot -p nucmer_mum_dsim_rcMAMBMCME -t postscript nucmer_mum_dsim.delta\n",
    "\n",
    "export GNUPLOT_PS_DIR=/home/nt365/pkg/gnuplot-4.6.0/term/PostScript \n",
    "gnuplot nucmer_mum_dsim_rcMAMBMCME.gp\n",
    "ps2pdf nucmer_mum_dsim_rcMAMBMCME.ps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4.29.2020   \n",
    "Remade mummer projects species folder, used references by the name indicated in \"-i\" from the following locations:\n",
    "Dana, Dere, Dyak, Dsim from my 3D-DNA output. Dana ref from /projects/genetics/ellison_lab/nicole/3D-DNA/3D-DNA_dana10/new_assembly\n",
    "Others from corresponding folder within /projects/genetics/ellison_lab/nicole/3D-DNA/\n",
    "Tiru assembled from corresponding species folder in /projects/genetics/ellison_lab/tiru_results\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Python script to extract scaffolds and rename as Muller element. Can reverse complement scaffolds as well if necessary."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<tt>**/projects/genetics/ellison_lab/nicole/mummer_projects/script/**<tt>\n",
    "\n",
    "instructions for \"extract_scaff_rename_muller.py\"\n",
    "Script is to rename HiC scaffolds from Drosophila species to appropriate Muller elements and reverse complement the scaffold if necessary.\n",
    "\n",
    "Be sure to input two output files, the first will print the Muller elements in the order they are found and the second will be the reordered Muller elements (A,B,C,D,E,F).\n",
    "\n",
    "in bash shell type:\n",
    "python extract_scaff_rename_muller.py -h \n",
    "\n",
    "Output will be \n",
    "Options:\n",
    "  -h, --help            show this help message and exit\n",
    "  -i FASTA              scaffolds.fasta\n",
    "  -o OUTFILE            output1.fasta\n",
    "  -t OUTFILE2, --o2=OUTFILE2\n",
    "                        output2.fasta - this will be the correctly ordered\n",
    "                        file\n",
    "  -a MA                 scaffold corresp. to Muller A\n",
    "  -b MB                 scaffold corresp. to Muller B\n",
    "  -c MC                 scaffold corresp. to Muller C\n",
    "  -d MD                 scaffold corresp. to Muller D\n",
    "  -e ME                 scaffold corresp. to Muller E\n",
    "  -f MF                 scaffold corresp. to Muller F\n",
    "  -u MAR, --rcMA=MAR    scaffold corresp. to Muller A - use to reverse\n",
    "                        complement scaff\n",
    "  -v MBR, --rcMB=MBR    scaffold corresp. to Muller B - use to reverse\n",
    "                        complement scaff\n",
    "  -w MCR, --rcMC=MCR    scaffold corresp. to Muller C - use to reverse\n",
    "                        complement scaff\n",
    "  -x MDR, --rcMD=MDR    scaffold corresp. to Muller D - use to reverse\n",
    "                        complement scaff\n",
    "  -y MER, --rcME=MER    scaffold corresp. to Muller E - use to reverse\n",
    "                        complement scaff\n",
    "  -z MFR, --rcMF=MFR    scaffold corresp. to Muller F - use to reverse\n",
    "                        complement scaff\n",
    "\n",
    "\n",
    "After running the script, input OUTFILE2 name into the <OUTFILE2> slot in the nucmer_slurm.sh script in the folder. Update prefix for nucmer output files, the folder for the SLURM output, and the nucmer parameters as necessary.  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#/projects/genetics/ellison_lab/nicole/mummer_projects/script/\n",
    "from optparse import OptionParser\n",
    "from Bio import SeqIO\n",
    "\n",
    "parser = OptionParser()\n",
    "parser.add_option(\"-i\", dest=\"fasta\", type=\"string\", action=\"store\", help = \"scaffolds.fasta\")\n",
    "parser.add_option(\"-o\", dest=\"outfile\", type = \"string\", action = \"store\", help = \"output1.fasta\") \n",
    "parser.add_option(\"-t\",\"--o2\", dest=\"outfile2\", type = \"string\", action = \"store\", help = \"output2.fasta - this will be the correctly ordered file\")\n",
    "parser.add_option(\"-a\", dest=\"MA\", type = \"string\", help = \"scaffold corresp. to Muller A\")\n",
    "parser.add_option(\"-b\", dest=\"MB\", type = \"string\", help = \"scaffold corresp. to Muller B\")\n",
    "parser.add_option(\"-c\", dest=\"MC\", type = \"string\", help = \"scaffold corresp. to Muller C\")\n",
    "parser.add_option(\"-d\", dest=\"MD\", type = \"string\", help = \"scaffold corresp. to Muller D\")\n",
    "parser.add_option(\"-e\", dest=\"ME\", type = \"string\", help = \"scaffold corresp. to Muller E\")\n",
    "parser.add_option(\"-f\", dest=\"MF\", type = \"string\", action = \"store\", help = \"scaffold corresp. to Muller F\")\n",
    "parser.add_option(\"-u\",\"--rcMA\", dest=\"MAR\", type = \"string\", action = \"store\", help = \"scaffold corresp. to Muller A - use to reverse complement scaff\")\n",
    "parser.add_option(\"-v\", \"--rcMB\", dest=\"MBR\", type = \"string\", action = \"store\", help = \"scaffold corresp. to Muller B - use to reverse complement scaff\")\n",
    "parser.add_option(\"-w\", \"--rcMC\", dest=\"MCR\", type = \"string\", action = \"store\", help = \"scaffold corresp. to Muller C - use to reverse complement scaff\")\n",
    "parser.add_option(\"-x\", \"--rcMD\", dest=\"MDR\", type = \"string\", action = \"store\", help = \"scaffold corresp. to Muller D - use to reverse complement scaff\")\n",
    "parser.add_option(\"-y\", \"--rcME\", dest=\"MER\", type = \"string\", action = \"store\", help = \"scaffold corresp. to Muller E - use to reverse complement scaff\")\n",
    "parser.add_option(\"-z\", \"--rcMF\", dest=\"MFR\", type = \"string\", action = \"store\", help = \"scaffold corresp. to Muller F - use to reverse complement scaff\")\n",
    "\n",
    "(options, args) = parser.parse_args()\n",
    "f = open(options.outfile, 'w')\n",
    "\n",
    "for record in SeqIO.parse(options.fasta, \"fasta\"):\n",
    "\tif record.id == options.MA:\n",
    "\t\tMA = record\t\n",
    "\t\tf.write(\">\" + \"Muller_A\" + '\\n' + str(MA.seq) + '\\n')\n",
    "\telif record.id == options.MB:\n",
    "\t\tMB = record\n",
    "\t\tf.write(\">\" + \"Muller_B\" + '\\n' + str(MB.seq) + '\\n')\n",
    "\telif record.id == options.MC:\n",
    "\t\tMC = record\n",
    "\t\tf.write(\">\" + \"Muller_C\" + '\\n' + str(MC.seq) + '\\n')\n",
    "\telif record.id == options.MD:\n",
    "\t\tMD = record\n",
    "\t\tf.write(\">\" + \"Muller_D\" + '\\n' + str(MD.seq) + '\\n')\n",
    "\telif record.id == options.ME:\n",
    "\t\tME = record\n",
    "\t\tf.write(\">\" + \"Muller_E\" + '\\n' + str(ME.seq) + '\\n')\n",
    "\telif record.id == options.MF:\n",
    "\t\tMF = record\n",
    "\t\tf.write(\">\" + \"Muller_F\" + '\\n' + str(MF.seq) + '\\n')\n",
    "\n",
    "for record in SeqIO.parse(options.fasta, \"fasta\"):\n",
    "        if record.id == options.MAR:\n",
    "                MAR = record\n",
    "                f.write(\">\" + \"Muller_A\" + '\\n' + str(MAR.seq.reverse_complement()) + '\\n')\n",
    "        elif record.id == options.MBR:\n",
    "                MBR = record\n",
    "                f.write(\">\" + \"Muller_B\" + '\\n' + str(MBR.seq.reverse_complement()) + '\\n')\n",
    "        elif record.id == options.MCR:\n",
    "                MCR = record\n",
    "                f.write(\">\" + \"Muller_C\" + '\\n' + str(MCR.seq.reverse_complement()) + '\\n')\n",
    "        elif record.id == options.MDR:\n",
    "                MDR = record\n",
    "                f.write(\">\" + \"Muller_D\" + '\\n' + str(MDR.seq.reverse_complement()) + '\\n')\n",
    "        elif record.id == options.MER:\n",
    "                MER = record\n",
    "                f.write(\">\" + \"Muller_E\" + '\\n' + str(MER.seq.reverse_complement()) + '\\n')\n",
    "        elif record.id == options.MFR:\n",
    "                MFR = record\n",
    "                f.write(\">\" + \"Muller_F\" + '\\n' + str(MFR.seq.reverse_complement()) + '\\n')\n",
    "\n",
    "f.close()\n",
    "\n",
    "r = open(options.outfile2, 'w')\n",
    "\n",
    "for record in SeqIO.parse(options.outfile, \"fasta\"):\n",
    "\tif record.id == \"Muller_A\":\n",
    "\t\tA = record\n",
    "\tif record.id == \"Muller_B\":\n",
    "\t\tB = record\n",
    "\tif record.id == \"Muller_C\":\n",
    "                C = record\n",
    "\tif record.id == \"Muller_D\":\n",
    "\t\tD = record\n",
    "\tif record.id == \"Muller_E\":\n",
    "                E = record\n",
    "\tif record.id == \"Muller_F\":\n",
    "                F = record\n",
    "\n",
    "\n",
    "\n",
    "r.write(\">\" + str(A.id) + '\\n' + str(A.seq) + '\\n')\n",
    "r.write(\">\" + str(B.id) + '\\n' + str(B.seq) + '\\n')\n",
    "r.write(\">\" + str(C.id) + '\\n' + str(C.seq) + '\\n')   \n",
    "r.write(\">\" + str(D.id) + '\\n' + str(D.seq) + '\\n')    \n",
    "r.write(\">\" + str(E.id) + '\\n' + str(E.seq) + '\\n')\n",
    "r.write(\">\" + str(F.id) + '\\n' + str(F.seq) + '\\n')\n",
    "r.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
